hw10/hw10_t01.py

```diff
@@ -0,0 +1,225 @@
+                        company_info["price"],
+                        company_info["P/E"],
+                        company_info["potential profit"],
+                    ) = await get_company_info(url, session, usd)
+
+                    evaluate_top10(company_info)
+    write_into_files()
+
+
+async def get_usd_currency_from_cbr(session: aiohttp.ClientSession) -> float:
+    today_date = dt.datetime.strftime(dt.datetime.today(), "%d/%m/%Y")
+    yesterday_date = dt.datetime.strftime(
+        dt.datetime.today() - dt.timedelta(days=1), "%d/%m/%Y"
+    )
+    usd_params = {
+        "date_req1": yesterday_date,
+        "date_req2": today_date,
+        "VAL_NM_RQ": "R01235",
+    }
+    async with session.get(
+        "http://www.cbr.ru/scripts/XML_dynamic.asp", params=usd_params
+    ) as resp:
+        html = await resp.text()
+        soup = BeautifulSoup(html, features="html.parser")
+        return float(soup.find_all("value")[-1].string.replace(",", "."))
+
+
+async def get_company_info(
+    url: str, session: aiohttp.ClientSession, usd: float
+) -> tuple:
+    async with session.get(url) as resp:
+        html = await resp.text()
+        soup = BeautifulSoup(html, features="html.parser")
+
+        current_value = float(
+            soup.find(
+                "span", attrs={"class": "price-section__current-value"}
+            ).string.replace(",", "")
+        )
+
+        code = list(
+            soup.find("span", attrs={"class": "price-section__category"}).strings
+        )[-2].strip(", ")
+
+        try:
+            p_e = float(
+                list(
+                    soup.find(
+                        "div", attrs={"class": "snapshot__header"}, string="P/E Ratio"
+                    ).parent.stripped_strings
+                )[0].replace(",", "")
+            )
+        except AttributeError:
+            p_e = None
+
+        try:
+            year_low = float(
+                list(
+                    soup.find(
+                        "div", attrs={"class": "snapshot__header"}, string="52 Week Low"
+                    ).parent.stripped_strings
+                )[0].replace(",", "")
+            )
+            year_high = float(
+                list(
+                    soup.find(
+                        "div",
+                        attrs={"class": "snapshot__header"},
+                        string="52 Week High",
+                    ).parent.stripped_strings
+                )[0].replace(",", "")
+            )
+            potential_profit = round((year_high - year_low) * usd, 2)
+        except AttributeError:
+            potential_profit = None
+
+        return code, current_value, p_e, potential_profit
+
+
+def evaluate_top10(company: dict) -> None:
+    evaluate_top10_price(company)
+    evaluate_top10_pe(company)
+    evaluate_top10_growth(company)
+    evaluate_top10_potential_profit(company)
+
+
+def evaluate_top10_price(company: dict) -> None:
```

Comment:
Could you describe the algorithm behind this code?  Have you implemented something like that before?

Reply:
At hw2 task01. Almost the same algorithm. At first, I wanted to collect info about every company, use 4 sorting filters, and take the top 10 for every requirement. It is O(NlogN). So I decided to do this instead. O(N)

Reply:
This code has O(n) for finding a place and O(n) for inserting into a list. I'd say it is way too big for finding a suitable location you can do better.

What is the best data structure which gives you 0(1) to find the smallest/biggest element and cheap to maintain?

Reply:
Might be heaps.  The interesting part is to integrate dict to a heap.

Reply:
Or maybe just `bisect` into a list? (moving all items in the list is slow though)

Reply:
> Might be heaps. The interesting part is to integrate dict to a heap.

You can use tuples `(-company["price"], company['code'], company)`

Tuples are compared from left to right.  You can use +- to change min/max,  the second element is guaranteed to be unique by task definition so we will never compare third one.